### ISSUE TITLE
[Deque Analysis] User is not warned of session timeout

### ISSUE THREAD

**Issue by tdonohue** [DSpace/dspace-angular#1173]
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary
Copied from Issue ID: 471071

Content times out without providing any warning to the user.

The current session has a time out of around 30 minutes and there is no warning to the user. 
The user gets logged out automatically and redirected to the login page.

RULE : 
If there is a session time limit, users MUST be warned before the session ends and MUST be given time to save their data and/or extend the session.

HOW TO FIX: 
Fix this issue by providing an accessible warning of the timeout and an accessible mechanism to do ONE of the following:
1. Turn off: Allow the user to turn off the time limit before encountering it.
2. Adjust: Allow the user to adjust the time limit before encountering it over a wide range that is at least ten times the length of the default setting.
3. Extend: Warn the user before time expires, give the user at least 20 seconds to extend the time limit with a simple action (for example, "press the spacebar"), and allow the user to extend the time limit at least ten times.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/dynamic-updates2/time-limits/session-timeout

## Recommended Fix

We may need to discuss the best way to implement this.  A few early ideas (off the top of my head)

* As the timeout is configurable, this may require letting the UI know the value of `jwt.login.token.expiration` (which is set to 1800000ms = 30mins by default).  We might just be able to make it available via `/api/config/properties/jwt.login.token.expiration`
* It seems like the UI may need to keep track of the last time it [refreshed the Auth token](https://github.com/DSpace/RestContract/blob/main/authentication.md#refresh-authentication)...and as the expiration time nears (perhaps within one minute) it should warn the user to click a button (or similar) to refresh the token.

**Comment by tdonohue** [comment]
@artlowel : I'd appreciate your feedback on this one (and we could bring to a future meeting if needed).  It might need more discussion, but I've added a few early ideas in the "Recommended Fix" section above.  If that sounds logical to you, we could move this forward into estimation...but, I'd be interested in other ideas as well.

**Comment by artlowel** [comment]
Currently there is an interceptor that will check that timestamp for every rest request you make. If the token is within 5 minutes (hardcoded) of expiring it gets automatically refreshed inside that interceptor without user interaction. If it has already expired you're redirected to the login page.

The problem with it is that it doesn't take caching in to account. If you're actively working but everything you need in that 5 minute window is already cached you won't trigger the refresh mechanism and you'll be logged out the next time you _do_ need something.

So I think we should take the redirect and refresh logic out of the auth interceptor, and make it simply add the auth header if the token exists, nothing more.

After the app initializes we start listening for token changes. If a token is stored, we set a timer to go off a configurable amount of time (e.g. 5 minutes by default) before it expires.

We also add a second timer that re-starts every time _any_ ngrx action occurs. Since these actions are used for every change to the state of the application they seem a sensible proxy for user activity without having to worry about tracking mouse movement or touch equivalents.

If this second timer completes within a configurable amount of time (e.g. 30 minutes by default) we show a modal saying:

"For security reasons, user sessions expire after x minutes of inactivity. Your session will expire soon. Would you like to extend it or log out?" 
 
If the user doesn't act within the configured amount of time we log them out automatically. Otherwise, we extend or log them out depending on their choice

If the token-validity-timer expires before the user-activity-timer does, we extend automatically in the background.

This will take an estimated **16 hours**

**Comment by tdonohue** [comment]
@artlowel : That all sounds reasonable to me.  However, we may wish to consider making the second timer ("user-activity-timer") have a default value based on the value of `jwt.login.token.expiration` (defaults to 30mins), as that's our real "session timeout".   

I just want to avoid a scenario where the backend has one timeout while the frontend has another -- so, ideally the frontend should read the config from the backend (however, if the backend config is unavailable, maybe there's a fallback on the frontend).

In any case, the direction here sounds fine...just noting we may want a way to share/sync the session timeout settings between frontend & backend.  I'll add the estimate & assign back to your team for development (whenever you are ready to start)

**Comment by artlowel** [comment]
While we could make them the same, they don't necessarily need to be. If we implement this as described, the backend's token expiration time is irrelevant for the UI. If the user is active when it expires it is extended automatically, and if the user is inactive before it expires, they're logged out automatically. The only thing that matters is the idle timeout, so you could configure that to be quite short (e.g. 15 minutes), because this is time we're sure the token isn't being used, while the token expiration time can be a lot longer (e.g. 2 hours), because this is the maximum "lease“ we give for the token regardless of use.

But I realize that's hard to convey in the docs, and we may get questions about why people can configure these similar properties separately for the UI and the backend. 

If we do decide to keep it in the backend, perhaps we should add the "grace period" (those 5 minutes) in the backend config as well so people don't have to configure these in separate places.

**Comment by tdonohue** [comment]
@artlowel : Good point, I hadn't thought of that in this way.  Ok, we can keep these separate. I think in the docs/comments we will need to clarify that the UI-level settings control when you are **automatically logged out** (from the UI), while the REST API settings are a global session expiration (it expires on all devices at the same time).

So, I think we can better describe this in documentation/comments.   Your approach seems fine & I agree that we don't need to implement a way to read the `jwt.login.token.expiration` setting from the backend.
